common/phal_L4/tim/tim.c

```diff
@@ -0,0 +1,166 @@
+/**
+ * @file tim.c
+ * @author Luke Oxley (lcoxley@purdue.edu)
+ * @brief Timer library meant for PWM input / output
+ * @version 0.1
+ * @date 2021-01-13
+ * 
+ * @copyright Copyright (c) 2021
+ * 
+ */
+#include "common/phal_L4/tim/tim.h"
+
+// prototypes
+bool enableTIMClk(TIM_TypeDef* timer);
+
+bool enableTIMClk(TIM_TypeDef* timer)
+{
+    // Enable Timer Clock 
+    if (timer == TIM1)
+    {
+        RCC->APB2ENR |= RCC_APB2ENR_TIM1EN;
+    }
+    else if (timer == TIM2)
+    {
+        RCC->APB1ENR1 |= RCC_APB1ENR1_TIM2EN;
+    }
+    else if (timer == TIM6)
+    {
+        RCC->APB1ENR1 |= RCC_APB1ENR1_TIM6EN;
+    }
+    else if (timer == TIM7)
+    {
+        RCC->APB1ENR1 |= RCC_APB1ENR1_TIM7EN;
+    }
+    else if (timer == TIM15)
+    {
+        RCC->APB2ENR |= RCC_APB2ENR_TIM15EN;
+    }
+    else if (timer == TIM16)
+    {
+        RCC->APB2ENR |= RCC_APB2ENR_TIM16EN;
+    }
+    else
+    {
+        return false;
+    }
+    return true;
+}
+
+// available on:
+// TIM1, TIM2: channel (1,2), (3,4)
+// TIM15 (1,2)
+// TIM2 is the only 32 bit resolution, currently 32 is not supported
+bool PHAL_initPWMIn(TIM_TypeDef* timer, uint32_t prescaler, TimerTriggerSelection_t trigger_select)
+{
+    if (!enableTIMClk(timer)) return false;
+    // can set the input prescaler (capture / events ICPSC)
+    // this is done if using interrupts and you want less of them
+
+    // configure prescaler
+    timer->PSC = prescaler - 1;
+    
+    // Update interrupt only on overflow
+    timer->CR1 |= TIM_CR1_URS;
+
+    // Select the trigger input TI1FP1
+    timer->SMCR |= (trigger_select << TIM_SMCR_TS_Pos) & TIM_SMCR_TS;
+    //TIM1->CR2 |= TIM_CR2_MMS_1;
+    // Set slave controller to reset mode (used for PWM input)
+    timer->SMCR |= TIM_SMCR_SMS_2;
+
+    // enable counter
+    timer->CR1 |= TIM_CR1_CEN;
+
+    return true;
+    // interrupt request through CC1IE in TIM_DIER
+}
+
+void PHAL_startTIM(TIM_TypeDef* timer)
+{
+    timer->CR1 |= TIM_CR1_CEN;
+}
+
+// initialize a capture compare channel
+bool PHAL_initPWMChannel(TIM_TypeDef* timer, TimerCCRegister_t chnl, TimerInputMode_t input_source, bool is_falling)
+{
+    // Input selection
+    if (chnl < 3)
+    {
+        timer->CCMR1 &= ~((input_source & TIM_CCMR1_CC1S) << (chnl - 1)*TIM_CCMR1_CC2S_Pos);
```

Comment:
It looks like you are trying to clear the previous input selection here, but you might be clearing the bits from the new selection only.
I think you want this: 
`timer->CCMR1 &= ~((TIM_CCMR1_CC1S) << (chnl - 1)*TIM_CCMR1_CC2S_Pos);`